File: music-lessons-2-java/app/src/main/java/com/mstoyanov/musiclessons/StudentsFragment.java
package com.mstoyanov.musiclessons;

import android.content.Intent;
import android.os.AsyncTask;
import android.os.Bundle;
import android.support.design.widget.FloatingActionButton;
import android.support.v4.app.Fragment;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ProgressBar;
import android.widget.TextView;

import com.mstoyanov.musiclessons.model.Student;
import com.mstoyanov.musiclessons.repository.AppDatabase;

import java.io.Serializable;
import java.lang.ref.WeakReference;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

public class StudentsFragment extends Fragment {
    private RecyclerView.Adapter adapter;
    private ProgressBar progressBar;
    private static List<Student> studentList;

    public static StudentsFragment create(int position) {
        StudentsFragment fragment = new StudentsFragment();
        Bundle args = new Bundle();
        args.putInt("POSITION", position);
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View rootView = inflater.inflate(R.layout.fragment_students, container, false);

        final TextView title = rootView.findViewById(R.id.heading);
        title.setText(R.string.students_label);

        final RecyclerView students = rootView.findViewById(R.id.students_list);
        studentList = new ArrayList<>();
        adapter = new StudentsAdapter(studentList, this);
        students.setAdapter(adapter);
        students.setLayoutManager(new LinearLayoutManager(getActivity()));

        progressBar = rootView.findViewById(R.id.progress_bar);
        progressBar.setIndeterminate(true);

        if (savedInstanceState == null) {
            new LoadStudents(progressBar, adapter).execute();
        } else {
            progressBar.setVisibility(View.GONE);
            studentList.addAll((List<Student>) savedInstanceState.getSerializable("STUDENTS"));
            adapter.notifyDataSetChanged();
        }

        FloatingActionButton button = rootView.findViewById(R.id.add_student);
        button.setOnClickListener(new View.OnClickListener() {
            public void onClick(View v) {
                startActivity(new Intent(getActivity(), AddStudentActivity.class));
            }
        });

        return rootView;
    }

    @Override
    public void onSaveInstanceState(Bundle state) {
        super.onSaveInstanceState(state);
        state.putSerializable("STUDENTS", (Serializable) studentList);
    }

    public void startProgressBar() {
        progressBar.setVisibility(View.VISIBLE);
    }

    public void stopProgressBar() {
        progressBar.setVisibility(View.GONE);
    }

    private static class LoadStudents extends AsyncTask<Long, Integer, List<Student>> {
        private final WeakReference<ProgressBar> progressBarWeakReference;
        private final WeakReference<RecyclerView.Adapter> adapterWeakReference;
        private static final AppDatabase DB = MusicLessonsApplication.getDB();

        LoadStudents(ProgressBar progressBar, RecyclerView.Adapter adapter) {
            progressBarWeakReference = new WeakReference<>(progressBar);
            adapterWeakReference = new WeakReference<>(adapter);
        }

        @Override
        protected List<Student> doInBackground(Long... params) {
            /*try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }*/
            return DB.getStudentDao().findAll();
        }

        @Override
        protected void onPostExecute(List<Student> result) {
            final ProgressBar progressBar = progressBarWeakReference.get();
            if (progressBar != null) progressBar.setVisibility(View.GONE);
            Collections.sort(result);
            studentList.addAll(result);
            final RecyclerView.Adapter adapter = adapterWeakReference.get();
            if (adapter != null) adapter.notifyDataSetChanged();
        }
    }
}

File: music-lessons-2-kotlin/app/src/main/java/com/mstoyanov/musiclessons/StudentsFragment.kt
package com.mstoyanov.musiclessons

import android.content.Intent
import android.os.AsyncTask
import android.os.Bundle
import android.support.design.widget.FloatingActionButton
import android.support.v4.app.Fragment
import android.support.v7.widget.LinearLayoutManager
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ProgressBar
import android.widget.TextView
import com.mstoyanov.musiclessons.model.Student
import java.io.Serializable
import java.lang.ref.WeakReference

class StudentsFragment : Fragment() {
    private lateinit var progressBar: ProgressBar
    private lateinit var adapter: StudentsAdapter
    private var students: MutableList<Student> = mutableListOf()

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val rootView = inflater.inflate(R.layout.fragment_students, container, false)

        val title = rootView.findViewById<TextView>(R.id.heading)
        title.setText(R.string.students_label)

        val recyclerView = rootView.findViewById<RecyclerView>(R.id.students_list)
        adapter = StudentsAdapter(students, this)
        recyclerView.adapter = adapter
        recyclerView.layoutManager = LinearLayoutManager(activity)

        progressBar = rootView.findViewById(R.id.progress_bar)
        progressBar.isIndeterminate = true

        if (savedInstanceState == null) {
            LoadStudents(this).execute()
        } else {
            progressBar.visibility = View.GONE
            students.addAll(savedInstanceState.getSerializable("STUDENTS") as MutableList<Student>)
            adapter.notifyDataSetChanged()
        }

        val button: FloatingActionButton = rootView.findViewById(R.id.add_student)
        button.setOnClickListener { startActivity(Intent(activity, AddStudentActivity::class.java)) }

        return rootView
    }

    override fun onSaveInstanceState(state: Bundle) {
        super.onSaveInstanceState(state)
        state.putSerializable("STUDENTS", students as Serializable)
    }

    fun startProgressBar() {
        progressBar.visibility = View.VISIBLE
    }

    fun stopProgressBar() {
        progressBar.visibility = View.GONE
    }

    companion object {

        fun create(position: Int): StudentsFragment {
            val fragment = StudentsFragment()
            val args = Bundle()
            args.putInt("POSITION", position)
            fragment.arguments = args
            return fragment
        }

        private class LoadStudents(context: StudentsFragment) : AsyncTask<Long, Int, MutableList<Student>>() {
            private val studentsFragmentWeakReference = WeakReference<StudentsFragment>(context)

            override fun doInBackground(vararg p0: Long?): MutableList<Student> {
                /*try {
                    Thread.sleep(1000)
                } catch (e: InterruptedException) {
                    e.printStackTrace()
                }*/
                return MusicLessonsApplication.db.studentDao.findAll()
            }

            override fun onPostExecute(result: MutableList<Student>) {
                studentsFragmentWeakReference.get()!!.progressBar.visibility = View.GONE
                result.sort()
                studentsFragmentWeakReference.get()!!.students.addAll(result)
                studentsFragmentWeakReference.get()!!.adapter.notifyDataSetChanged()
            }
        }
    }
}